### ISSUE TITLE
For an ML Program, extension must be .mlpackage (not .mlmodel)

### ISSUE THREAD

**Issue by ShreshtMG** [ultralytics/ultralytics#2088]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Export

### Bug

Ultralytics YOLOv8.0.81 🚀 Python-3.9.16 torch-2.0.0+cu118 CPU
Model summary (fused): 168 layers, 3014033 parameters, 0 gradients, 8.1 GFLOPs

PyTorch: starting from /content/drive/MyDrive/ColabNotebooks/bestGTSRB.pt with input shape (1, 3, 640, 640) BCHW and output shape(s) (1, 47, 8400) (5.9 MB)

CoreML: starting export with coremltools 6.3.0...
WARNING:coremltools:Tuple detected at graph output. This will be flattened in the converted model.
Converting PyTorch Frontend ==> MIL Ops: 100%|█████████▉| 523/525 [00:00<00:00, 3718.19 ops/s]
Running MIL frontend_pytorch pipeline: 100%|██████████| 5/5 [00:00<00:00, 202.43 passes/s]
Running MIL default pipeline: 100%|██████████| 56/56 [00:00<00:00, 141.97 passes/s]
Running MIL backend_neuralnetwork pipeline: 100%|██████████| 8/8 [00:00<00:00, 498.59 passes/s]
Translating MIL ==> NeuralNetwork Ops:  89%|████████▉ | 568/639 [00:00<00:00, 1626.08 ops/s]WARNING:coremltools:Const anchor_points was already added.
Translating MIL ==> NeuralNetwork Ops: 100%|██████████| 639/639 [00:00<00:00, 1769.38 ops/s]
CoreML Pipeline: starting pipeline with coremltools 6.3.0...
CoreML Pipeline: pipeline success
CoreML: export failure ❌ 6.4s: For an ML Program, extension must be .mlpackage (not .mlmodel)
[]

### Additional

Cannot export model to CoreML format with nms

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@ShreshtMG, thank you for reporting the issue! It seems like you are encountering an error when you are trying to export the model to CoreML format. The error message suggests that the extension of the file must be ".mlpackage" instead of ".mlmodel" for your machine learning program. Can you try changing the file extension to ".mlpackage" and try exporting it again? Also, make sure you have installed [coremltools](https://pypi.org/project/coremltools/) version 4.1.0 or later through `pip`, which is required for CoreML export capability. Let us know if this resolves the problem.

**Comment by ShreshtMG** [comment]
Hi @glenn-jocher , thank you very much for your reply. Yes, you are right, I am exporting the model to CoreML format. However, the problem is, while exporting the saved PyTorch model to CoreML format, there are no issues, but when I change the nsm argument to True in the export function, the model is failing to export. Please have a look at the screenshot below.

![image](https://user-images.githubusercontent.com/125672809/232599599-02cdd480-aa43-4c99-ac9a-ea63fbf24834.png)

Regarding what you have mentioned, do I need to export the .mlmodel to .mlpackage or .pt to .mlpackage? I tried either ways, unfortunately nothing worked

**Comment by glenn-jocher** [comment]
@ShreshtMG, thank you for the clarification. Based on the screenshot, it looks like the error message is related to a mismatch between the input and output shapes of the model. This suggests that the inputs and outputs of the model are not of the correct dimensions.

Regarding the export format, please note that `.mlpackage` is the recommended format for CoreML models. It seems like you have already tried exporting the model in the correct format, but the export function with `nms` is still not functioning correctly.

Can you provide more information about the model's architecture and dimensions of the input and output tensors? This can help us better understand the problem and provide more specific guidance.

**Comment by ShreshtMG** [comment]
Hi @glenn-jocher, thank you very your response and clarification. You got straight to the point. I hope the screenshot below answers the question, as it is said that a picture is worth thousands of words. 

![image](https://user-images.githubusercontent.com/125672809/232670076-acba08c2-85b4-49be-b4cc-588bd675bc7e.png)

**Comment by glenn-jocher** [comment]
@ShreshtMG yes I see now. This is due to an Apple change starting in coremltools 6.3.0. As a quick fix you can downgrade to 6.2.0 and your command will run correctly. 

We are working on a permanent fix in https://github.com/ultralytics/ultralytics/pull/1791

**Comment by ShreshtMG** [comment]
Thanks for your response @glenn-jocher. Are there any resources on how to post-processing the output to get bounding boxes, and score?

**Comment by ShreshtMG** [comment]
Also, wanted to bring to your notice that when I run the benchmark on my model, it only exports the model. However, it could not perform inference on it.

<img width="1091" alt="Screenshot 2023-04-18 at 15 39 32" src="https://user-images.githubusercontent.com/125672809/232812169-314dc4ba-2765-4968-b12b-f7ca119e3150.png">

**Comment by glenn-jocher** [comment]
@ShreshtMG, post-processing the model output can vary depending on the specific use case and task. In general, the output of a YOLOv8 model trained for object detection includes information about the bounding boxes of the detected objects and their corresponding confidence scores.

A common approach to post-processing the output is Non-Maximum Suppression (NMS), which is used to eliminate overlapping bounding boxes that may be predicted for the same object. This helps reduce the number of false positives and improve the accuracy of object detection.

Regarding the benchmark issue you reported, it seems like the benchmark script is having trouble loading the exported CoreML model. This might be due to a mismatch between the model architecture and the input/output shapes. You can verify that the exported model is valid by testing it on some sample data outside of the benchmark script. If you continue to have issues, please provide more details about the error message or other relevant information.

**Comment by philipperemy** [comment]
Long story short. How to fix this:

```
pip install "coremltools<6.3.0"
```

**Comment by glenn-jocher** [comment]
@philipperemy sure, here is a brief description. 

The error that you are experiencing during the export of your PyTorch model to CoreML format is due to a change in coremltools version 6.3.0 which causes compatibility issues with your model. A quick solution to this problem is to rollback to the previous version of coremltools, which can be done by using the pip command `pip install "coremltools<6.3.0"`. This will install a compatible version of coremltools and allow you to successfully export your model to CoreML format.

**Comment by mwyborski** [comment]
I was able to fix the error for me by replacing

    f = self.file.with_suffix('.mlmodel')

with

    f = self.file.with_suffix('.mlpackage')

in the function **export_coreml()** in the file **ultralytics/engine/exporter.py**

https://github.com/ultralytics/ultralytics/blob/ed25db94268fe0e65030f0ddc55e6180e39ffe81/ultralytics/engine/exporter.py#L464C46-L464C46

**Comment by glenn-jocher** [comment]
@mwyborski thank you for your suggestion. This change makes sense given that starting from coremltools 6.3.0, the `.mlpackage` format is recommended for CoreML models. We will consider your suggested change and see if it can be incorporated into our future updates. Please stay tuned for updates in the future versions of the YOLOv8 codebase. Your contribution is greatly appreciated!

**Comment by glenn-jocher** [comment]
@philipperemy @mwyborski @ShreshtMG great news 😃! Your original issue may now be resolved ✅ in PR #4043. This PR is not merged yet, but it will add direct support for *.mlpackage export where possible. 

To get this update once the PR is merged:
- **[Git](https://github.com/ultralytics/ultralytics)** – Run `git pull` from within your `ultralytics/` directory or run `git clone https://github.com/ultralytics/ultralytics` again
- **[Pip](https://pypi.org/project/ultralytics/)** – Update with `pip install -U ultralytics`
- **[Notebooks](https://github.com/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb)** – Check out the updated notebooks  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/ultralytics)** – Run `sudo docker pull ultralytics/ultralytics:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and letting us know. Please confirm if this update fixes the issue for you, and don't hesitate to report any other **issues** you find or **feature requests** you may have. Happy training with YOLOv8 🚀!

**Comment by kipkoalex** [comment]
Hello @glenn-jocher !
First of all I want to thank you for such incredible product and for all of your support!
I tried the code from mentioned PR, and it seems like there are a couple of small bugs in `/ultralytics/engine/exporter.py` happening when ones is trying to export `coreml` model with `nms=True` flag:
1. At line 865 we need to change
`model = ct.models.MLModel(spec)`
to 
`model = ct.models.MLModel(spec, weights_dir=model.weights_dir)`
Since weights_dir is required parameter when we're building a model from spec. Otherwise the code is failing.
2. The same needs to be added at line 924:
`model = ct.models.MLModel(pipeline.spec, weights_dir=model.weights_dir)` - if we not specify `weights_dir` here then resulting model will be saved without them.
Let me know if I'm wrong :)

**Comment by glenn-jocher** [comment]
@kipkoalex hi there,

Thank you for your appreciation and for pointing out these potential issues when exporting a CoreML model with the nms=True flag.

Regarding your first point, your suggestion to include 'weights_dir' as a necessary parameter when building a model from a spec makes sense as it ensures the model is linked to the correct weights.

Concerning your second point, including 'weights_dir' in that line of code makes logical sense for the same reasons – it should ensure that the exported model includes the necessary weight information.

Your contribution is greatly appreciated! We'll look into these observations and consider incorporating your suggestions into our codebase. Thank you again for your feedback!

**Comment by philipperemy** [comment]
@glenn-jocher awesome!

**Comment by glenn-jocher** [comment]
@philipperemy I'm glad you found the updates helpful! Thank you for your feedback and for using YOLOv8. If you have any more questions or run into any more issues, feel free to let us know. Happy training!
